src/mozanalysis/stats/summarize_samples.py

```diff
@@ -0,0 +1,195 @@
+    parameters; in both cases `samples` relate to data from one branch
+    of an experiment.
+
+    Some example use-cases:
+    - When building a time series for one metric, each column's
+        samples are derived from data for a different analysis window.
+    - When bootstrapping several quantiles simultaneously (with a
+        `stat_fn` that returns several values), each column's samples
+        come from bootstrapping a different quantile. N.B. in this
+        example, the quantiles being bootstrapped over are unrelated to
+        the `quantiles` argument of this function.
+
+    Args:
+        samples (pandas.DataFrame): Each column contains a set of
+            samples over which to compute the means and quantiles.
+        quantiles (list, optional): The quantiles to compute - a good
+            reason to override the defaults would be when Bonferroni
+            corrections are required.
+
+    Returns a pandas DataFrame; the columns contain the stringified
+    `quantiles` plus `'mean'`. The index matches the columns of
+    `samples`.
+    """
+    return samples.agg(summarize_one_branch_samples, quantiles=quantiles).T
+
+
+def summarize_joint_samples(focus, reference, quantiles=default_quantiles):
+    """Return descriptive statistics for uplifts.
+
+    The intended use case of this function is to compare a 'focus'
+    experiment branch to a 'reference' experiment branch (e.g. the
+    control). Samples from each branch are combined pairwise; these
+    pairs are considered to be samples from the joint probability
+    distribution (JPD). We compute various quantities from the JPD:
+    - We compute summary statistics for the distribution over relative
+        uplifts `focus / reference - 1`
+    - We compute summary statistics for the distribution over absolute
+        uplifts `focus - reference`
+    - We compute a summary statistic for the distribution over the L1
+        norm of absolute uplifts `abs(focus - reference)`
+    - We compute the fraction of probability mass in the region
+        `focus > reference`, which in a Bayesian context may be
+        interpreted as the probability that the ground truth model
+        parameter is larger for the focus branch than the reference
+        branch.
+
+    `focus` and `reference` are samples from distributions; each is the
+    same format that would be supplied to `summarize_one_branch_samples`
+    when analyzing the branches independently.
+
+    Args:
+        focus (pandas.Series): Bootstrapped samples or samples of a
+            model parameter for a branch of an experiment.
+        reference (pandas.Series): The same quantity, calculated for a
+            different branch (typically the control).
+        quantiles (list, optional): The quantiles to compute - a good
+            reason to override the defaults would be when Bonferroni
+            corrections are required.
+
+    Returns a pandas Series containing the following data:
```

Comment:
> avoid encoding quantiles in strings

Let's start with [this simpler conversation](https://github.com/mozilla/mozanalysis/pull/44/files/af80b1ce1476f54ddfa5197cd38dae78860e64f8#r288797470)

> avoid abbreviations

Would you prefer `relative_uplift_*` and `absolute_uplift_*`? What about `maximum_absolute_difference_*`? Bear in mind that the longer the label is, the more likely that it gets truncated with the value missing (so we can't distinguish `relative_uplift_0.005` from `relative_uplift_0.995`)... 

> using a Series here doesn't feel natural to me

I was tossing up between a dict or Series. Here are the features that drove me to use a Series:
* I want values to be ordered (grouping related quantities together and ordering withing those groups) and keyed
* I want the keys to be flexible (an arbitrary list of quantiles might be supplied)
* Batch mode will smash a bunch of these return values together, and the user should be able to conveniently address the collection of results either by row or column (i.e. get all quantiles for one set of samples, or get one quantile across all sets of samples). To me this really smells like a DataFrame.

Does that go some way to convince you that a Series is better here than an OrderedDict, dict or custom type? 

> Am I right that most users most of the time will probably care about the (0.025, 0.975) interval for a 95% CI?

Quite possibly? Personally I like to present 90% and 99% CIs/probability bands, to communicate the fuzziness of results and discourage premature "binarification" of outcomes; but this might not catch on and the function should default to what the team would choose to use if they had a long and informed discussion. Personally I would like to continue using 90% and 99% CIs/probability bands (unless there's a reason why they're problematic) and I would ideally like this function to support that, though it does not need to be the default.

Also I don't know whether people care more about absolute uplift or relative uplift; personally I've found relative uplift to be more useful most of the time, but there have been cases when absolute uplift has added perspective (e.g. when you forget that you're looking at a 0.01% absolute change on a very low baseline and you needed a 10% absolute change). So I threw both in here.

> I wonder if it doesn't make sense to spec these in terms of the interval width rather than the individual quantiles.

When I apply this function to samples from a posterior probability distribution, I like communicating to stakeholders that "There is a 5% chance that the metric will drop by more than `res.loc[rel_uplift_0.05]`" so I'd like the labels to reference the individual quantiles. If you think it makes more sense to specify interval width and have it spit out quantile labels then sure, I'm not overly fussed. I defaulted to having the argument more directly reflect the labels, but it does add a probably-unnecessary degree of freedom. Have your thoughts wandered to anywhere stronger than "wonder"?

Reply:
>> avoid encoding quantiles in strings

> Let's start with this simpler conversation

Oh, no, I do mean _in_ strings and not _as_ strings (which is not an obvious distinction and I apologize) -- I was trying to imagine how I'd prefer to do this and I had the same concerns you identify about the ergonomics of indexing a dictionary with a float, since floats are lies and we can't have nice things, so representing decimal values as strings is `this_is_fine.gif`. Parsing them out of texty labels is the only thing that seems annoying.

> I was tossing up between a dict or Series. Here are the features that drove me to use a Series:
> * I want values to be ordered (grouping related quantities together and ordering withing those groups) and keyed
> * I want the keys to be flexible (an arbitrary list of quantiles might be supplied)
> * Batch mode will smash a bunch of these return values together, and the user should be able to conveniently address the collection of results either by row or column (i.e. get all quantiles for one set of samples, or get one quantile across all sets of samples). To me this really smells like a DataFrame.
> Does that go some way to convince you that a Series is better here than an OrderedDict, dict or custom type?

I think I agree with your design goals; I'm tripping over the flat indexing scheme, since it feels like we "really" have a handful of different kinds of numbers rather than a single collection of numbers. One consequence is that I don't agree that the current elements of the omnibus Series have a natural total ordering and I don't really understand why you'd want to maintain one, except for display, and you can provide that with a nice `__str__` method.

I think a useful concrete alternative would be an object including `relative_uplift` and `absolute_uplift` members of type Series, indexed by the string "expectation" or str(quantile), which I think meets the latter two goals. The batching would upzone those into one DataFrame each (and a Series for each of the scalar values you want to include).

> Have your thoughts wandered to anywhere stronger than "wonder"?

I've gotten used to the idea now and you convinced me that there's a natural use case for specifying them as quantiles and I'm convinced I can serialize that in a useful way; I think specifying as quantiles makes sense.

Reply:
> Parsing them out of texty labels is the only thing that seems annoying.

I suppose my tastes have been warped by matching records based on hashes! I was not imagining that people would _parse_ them out of texty labels; I expected people to (magically?) know what the semantic values were and use those semantic values to recreate the relevant texty label...

As in, here, we would the expect that whoever called `summarize_joint_samples_batch` would know what quantiles were used, and could recreate the `'rel_uplift_{}'.format(q)` (yes we would need to make this more convenient if we go with the current design). I guess the deep question is "can/should the columns have labels that tell the caller everything they need to know about how to treat the data in that column?" and I'm not convinced the answer is "yes" - because for the quantiles the caller needs to know what quantile it is (a float) and what role it plays in the analysis (is it a CI? is it one of probability bands, and if so, which?). If the caller can't infer everything from the labels, then they have to already store some information about what labels they expect to be in the output, and so this label-reconstruction approach is less limiting.

So: yes, it's a bit annoying, but it shouldn't have architectural implications and therefore _might_ be the kind of thing you get over quickly?

> I'm tripping over the flat indexing scheme, since it feels like we "really" have a handful of different kinds of numbers rather than a single collection of numbers.

Follow me down the slippery slope: wouldn't it be nice if the columns were a multi-index, where the highest level labels are `['relative_uplift', 'absolute_uplift', 'max_abs_diff', 'prob_win']`, and all but `'prob_win'` have a lower level label per quantile value. Yes the highest level of the index doesn't have an important order, but overall the DataFrame's columns do fall into groups and within each group there is an ordering. This nicely separates the collections of numbers and allows you to address one collection at a time.

But multiindexes are awkward/fiddly and don't offer much over a flattened index, other than aesthetics. If you're going to address each column of data by its label, then maybe it's not the end of the world if you carry around some extra columns - as in if your function `plot_absolute_uplift_with_error_bars()` already needs to know the labels of the columns with relevant data, perhaps it's not atrocious if the data object also happens to have a bunch of columns about relative uplift that go unused?

The other aspect I like about having just one DataFrame here is that every column shares an index. Indexes will eventually represent one of a few different dimensions (multi-output `stat_fn`, thresholds, points in time that the analysis window starts); IMO it's nice to know that the `relative_uplift` and `absolute_uplift` objects implicitly share an index.

I think this is a situation where both sides have strengths and weaknesses; if you consider these issues and decide that an extra layer of data structure will make downstream use easier/neater/safer, then sure, we can stick these in a `dict` (or an `attr` class)

Reply:
I've changed from 90% and 99% CIs to 95% and 99% because most people here use 95% CIs (should I change to 95% and 99.5%?).

My assertion that text labels are not too awful hinges on mozreport being aware of what the input quantiles were - i.e. it will call the stats functions and therefore be aware of what quantiles were fed in (and look up the defaults if none were specified), and therefore be able to reconstruct the labels it needs. Is this a reasonable assumption?